packages/avatar/contracts/nft-collection/NFTCollection.sol

```diff
@@ -0,0 +1,1034 @@
+     */
+    function mint(
+        address wallet,
+        uint256 amount,
+        uint256 signatureId,
+        bytes calldata signature
+    ) external whenNotPaused nonReentrant {
+        NFTCollectionStorage storage $ = _getNFTCollectionStorage();
+        uint256 waveIndex = $.waveData.length;
+        if (waveIndex == 0) {
+            revert ContractNotConfigured();
+        }
+        if (_msgSender() != address($.allowedToExecuteMint)) {
+            revert ERC721InvalidSender(_msgSender());
+        }
+        _checkAndSetMintSignature(wallet, signatureId, signature);
+        WaveData storage waveData = $.waveData[waveIndex - 1];
+        _doMint(waveData, wallet, amount);
+    }
+
+    /**
+     * @notice token minting function on a certain wave. Price is set by wave and is paid in tokens denoted
+     *         by the allowedToExecuteMint contract
+     * @param wallet minting wallet
+     * @param amount number of token to mint
+     * @param waveIndex the index of the wave used to mint
+     * @param signatureId signing signature ID
+     * @param signature signing signature value
+     */
+    function waveMint(
+        address wallet,
+        uint256 amount,
+        uint256 waveIndex,
+        uint256 signatureId,
+        bytes calldata signature
+    ) external whenNotPaused nonReentrant {
+        NFTCollectionStorage storage $ = _getNFTCollectionStorage();
+        if ($.waveData.length == 0) {
+            revert ContractNotConfigured();
+        }
+        if (_msgSender() != address($.allowedToExecuteMint)) {
+            revert ERC721InvalidSender(_msgSender());
+        }
+        _checkAndSetWaveMintSignature(wallet, amount, waveIndex, signatureId, signature);
+        WaveData storage waveData = _getWaveData(waveIndex);
+        _doMint(waveData, wallet, amount);
+    }
+
+    /**
+     * @notice function to setup wave parameters. A wave is defined as a combination of allowed number tokens to be
+     *         minted in total, per wallet and minting price
+     * @param waveIndex the index of the wave to be canceled
+     */
+    function cancelWave(uint256 waveIndex) external onlyOwner {
+        NFTCollectionStorage storage $ = _getNFTCollectionStorage();
+        /// @dev don't use _getWaveData, we don't want to cancel the last wave by mistake
+        if (waveIndex >= $.waveData.length) {
+            revert ContractNotConfigured();
+        }
+        $.waveData[waveIndex].waveMaxTokensOverall = 0;
+    }
+
+    /**
+     * @notice batch minting function, used by owner to airdrop directly to users.
+     * @dev this methods takes a list of destination wallets and can only be used by the owner of the contract
+     * @param waveIndex the index of the wave used to mint
+     * @param wallets list of destination wallets and amounts
+     */
+    function batchMint(uint256 waveIndex, BatchMintingData[] calldata wallets) external whenNotPaused onlyOwner {
```

Comment:
I feel like the owner should be able to batch mint tokens without setting up a wave.
If we want to just mint internally we need to setup a wave for 1000 tokens to mint 1000 tokens all for ourselves.

Is there a reason why we would do it that way?

Reply:
I also think batch minting should not be so restrictive, we may want to be able to mint all to same wallet etc.
I think we should just check if we are not minting more than the total supply of the collection.